### PR TITLE
Phase2 OT Rings and Navigation fixed for tk geometry v4021

### PR DESCRIPTION
**Pull request by ebrondol**
This PR takes care of the number of rings in the new layout for the phase 2 tracker - the so-called V4021. Before, the number of rings in the OT was fixed to 15, now the size can change between rings. Moreover, this PR also fix the Navigation School in this layout: the logic in the forward layers ordering was not suitable anymore. A fast fix has been applied, but it can also be modify with something better when the reconstruction and the layout will be definitive.
The new navigation map can be found [d4.pdf](https://github.com/cms-sw/cmssw/files/528076/d4.pdf) and it does appear as it should. In order to produce it by yourself, the following analyzer can be used: `RecoTracker/TkNavigation/test/NavigationSchoolAnalyzer_cfg.py`.

The others geometry should stay unchanged. In order to facilitate the comparison, I have already produced the MTV results with ttbar sample for [run2](http://ebrondol.web.cern.ch/ebrondol/Phase2Tracking/PR16207/20161013_Run2_phase2_fixNavigation/), [phase 1](http://ebrondol.web.cern.ch/ebrondol/Phase2Tracking/PR16207/20161013_Phase1_phase2_fixNavigation/) and [phase 2 - D1 geometry](http://ebrondol.web.cern.ch/ebrondol/Phase2Tracking/PR16207/20161013_Phase2D1_phase2_fixNavigation/).

More fixes are needed in order to get the new phase2 geometry ready for performance, but with this PR it should be possible to create the entire D4 WF running (@boudoul). So, please consider this PR as urgent. Thanks.

@kpedro88  @makortel @atricomi @venturia

### CODE REVIEW

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by ebrondol** [comment]
With this last commit, I update the pixel cluster shape filter to be off for phase2 as [discussed](https://indico.cern.ch/event/578066/).
I have produced privately the results for extended muons for [D1](http://ebrondol.web.cern.ch/ebrondol/FixOT/20161019_Phase2D1_FixPhase2NavD4_TTbar_doPixelShapeCut_complete/) and [D4](http://ebrondol.web.cern.ch/ebrondol/FixOT/20161017_FixPhase2NavD4_FourMuons_doPixelShapeCut_complete/). For D1, we see better efficiency (especially at low pt) and increment in fakes is not as important, instead for D4 the eff in the barrel region is completely restored. In summary, very good news.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
urgent

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
+1

For #16207 30961bd8d3702b8b3833b4d3781fc259cea82a0c

Phase 2 outer tracker ring and navigation fixes, plus disabling the Phase 2 Pixel shape cut since these clusters are not yet correct.

The code changes are satisfactory. Jenkins tests against baseline CMSSW_8_1_X_2016-10-21-1100 show the expected numerous small differences in Phase 2 workflows due to the increased inner tracker efficiency caused by disabling the shape cut. A test of workflow 21200.0_FourMuPt1_200+FourMuPt_1_200_pythia8_2023D4 shows it runs successfully and has no difference in RECO event content compared with baseline CMSSW_8_1_X_2016-10-12-2300.

File: RecoTracker/TkDetLayers/src/Phase2OTEndcapLayer.cc [hunk 1]
```diff
@@ -248,19 +250,19 @@ Phase2OTEndcapLayer::ringIndicesByCrossingProximity(const TrajectoryStateOnSurfa
   
   Crossing myXing(  startPos, startDir, rho, propDir );
 
-  GlobalPoint   ringCrossings[NOTECRINGS];
+  std::vector<GlobalPoint> ringCrossings;
   // vector<GlobalVector>  ringXDirections;
```
Comment (davidlange6) [inline RecoTracker/TkDetLayers/src/Phase2OTEndcapLayer.cc]: @ebrondol - a ringCrossings.reserve(theRingSize) would be of benefit here

**Comment by ebrondol** [comment]
@davidlange6 @slava77 
please let me know if these fixes can be considered acceptable.

**Comment by davidlange6** [comment]
please test
(looks good to me)

**Comment by cvuosalo** [comment]
+1

For #16207 808099a17667fd0bd667e9af7edd42441536718b

Second approval, after very small code fix. The code revision is satisfactory, and the new Jenkins tests show the same expected changes as before.

**Comment by davidlange6** [comment]
+1

**Pull request merged**